crates/ruff_formatter/src/builders.rs

```diff
@@ -276,8 +276,60 @@ impl std::fmt::Debug for StaticText {
     }
 }
 
-/// Creates a text from a dynamic string and a range of the input source
-pub fn dynamic_text(text: &str, position: TextSize) -> DynamicText {
+/// Creates a source map entry from the passed source `position` to the position in the formatted output.
+///
+/// ## Examples
+///
+/// ```
+/// /// ```
+/// use ruff_formatter::format;
+/// use ruff_formatter::prelude::*;
+///
+/// # fn main() -> FormatResult<()> {
+/// // the tab must be encoded as \\t to not literally print a tab character ("Hello{tab}World" vs "Hello\tWorld")
+/// use ruff_text_size::TextSize;
+/// use ruff_formatter::SourceMarker;
+///
+///
+/// let elements = format!(SimpleFormatContext::default(), [
+///     source_position(TextSize::new(0)),
+///     text("\"Hello "),
+///     source_position(TextSize::new(8)),
+///     text("'Ruff"),
+///     source_position(TextSize::new(12)),
+///     text("'\"")
+/// ])?;
+///
+/// let printed = elements.print()?;
+///
+/// assert_eq!(printed.as_code(), r#""Hello 'Ruff'""#);
+/// assert_eq!(printed.sourcemap(), [
+///     SourceMarker { source: TextSize::new(0), dest: TextSize::new(0) },
+///     SourceMarker { source: TextSize::new(0), dest: TextSize::new(7) },
```

Comment:
There are two markers because we need to know where the position for a source position starts and ends in the formatted document.